src/System.IO.FileSystem/src/System/IO/Win32FileStream.cs

```diff
@@ -1291,19 +1305,45 @@ private Task WriteInternalAsync(byte[] array, int offset, int numBytes, Cancella
                 _readLen = 0;
             }
 
-            int n = _bufferSize - _writePos;
-            if (numBytes <= n)
+            int remainingBuffer = _bufferSize - _writePos;
+
+            // if the incoming write fits in the remaining buffer, treat
+            // as synchronous since memcopy is so fast
+            if (numBytes <= remainingBuffer)
             {
                 if (_writePos == 0) _buffer = new byte[_bufferSize];
                 Buffer.BlockCopy(array, offset, _buffer, _writePos, numBytes);
                 _writePos += numBytes;
-                
+
                 return Task.CompletedTask;
             }
 
-            if (_writePos > 0)
-                FlushWrite(false);
+            // if the incoming write fits in the current buffer + next buffer,
+            // only perform a single write to flush the existing buffer and fill
+            // the next buffer with the remainder
+            if (numBytes <= (_bufferSize + remainingBuffer))
+            {
+                if (_writePos == 0) _buffer = new byte[_bufferSize];
+
+                Buffer.BlockCopy(array, offset, _buffer, _writePos, remainingBuffer);
+
+                byte[] oldBuffer = _buffer;
+                _buffer = new byte[_bufferSize];
```

Comment:
Actually, I think I may be making this more complicated than it needs to be (Friday afternoon brain):
- If the write is truly asynchronous, the caller can't reuse the buffer until the Task is completed.  So in general they can't assume the buffer is reusable until then.
- Given that, worst case we can do something akin to what you've done below, where you just hold onto the user's buffer until it's time to actually write it out.  This is essentially what the base Stream does when there are other async operations in progress, chaining them all together.

As such, what you have is probably fine, except that I'm not sure it's worth the extra buffer allocation for this case.  What if we just said:
- When the data fits entirely in the current buffer, write it to the buffer and return synchronously.
- When it overflows the current buffer, write out the current buffer (if there is one) and then write out the user's buffer.

Essentially I think that's exactly what you have, except just deleting this additional case you added.

Thoughts?